Doc/reference/compound_stmts.rst

```diff
@@ -157,17 +157,16 @@ The :keyword:`for` statement is used to iterate over the elements of a sequence
    for_stmt: "for" `target_list` "in" `starred_list` ":" `suite`
            : ["else" ":" `suite`]
 
-The expression list is evaluated once; it should yield an iterable object.  An
-iterator is created for the result of the ``starred_list``.  The expression
-list can contain starred elements (``*x, *y``) that will be unpacked in the
-final iterator (as when constructing a ``tuple`` or ``list`` literal). The
-suite is then executed once for each item provided by the iterator, in the
-order returned by the iterator.  Each item in turn is assigned to the target
-list using the standard rules for assignments (see :ref:`assignment`), and then
-the suite is executed.  When the items are exhausted (which is immediately when
-the sequence is empty or an iterator raises a :exc:`StopIteration` exception),
-the suite in the :keyword:`!else` clause, if present, is executed, and the loop
-terminates.
+The starred list expression is evaluated once; it should yield an iterable object.
+If the expression is a sequence, it can contain starred elements (``*x, *y``)
+that will be unpacked, as when constructing a ``tuple`` or ``list`` literal.
+An iterator is created for the resulting iterable.  The first item provided
+by the iterator is then assigned to the target list using the standard
+rules for assignments (see :ref:`assignment`), and the suite is executed - this
+repeats for each item provided by the iterator.  When the items are exhausted
+(which is immediately when the sequence is empty or an iterator raises
+a :exc:`StopIteration` exception), the suite in the :keyword:`!else` clause,
```

Comment:
```suggestion
repeats for each item provided by the iterator.  When the iterator is exhausted,
the suite in the :keyword:`!else` clause,
```
We don't have to explain here how iterators work, and the current text is imprecise for classes using the old-style iteration protocol.

Perhaps we can link to the `iterator` glossary entry though.

Reply:
As I remember from experiments a few years ago, old-style iteration is done via a new-style iterator that serves as an adaptor.  The looping code only sees the adaptor and does not know that it wraps a get_item that raises IndexError.  But I agree that we do not need to explain 'exhausted' again here.  Let's link the first uses of both `iterable` and `iterator`.

Reply:
Agree, cross-references to the definitions of `iterable` and `iterator` would be good.

Reply:
done